File: cpp_src/puppypi_mock/fake_motion_server.cpp
// Fake Motion Server 实现
//
// 仿真 PuppyPi 运动平台
#include "puppypi_mock/fake_motion_server.h"

#include <chrono>
#include <cctype>
#include <cmath>
#include <functional>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace puppypi_mock {

FakeMotionServerNode::FakeMotionServerNode()
    : rclcpp::Node("fake_motion_server") {
  // 运动状态初始化
  standing_ = false;
  moving_ = false;
  controllable_ = false;
  current_posture_ = "SIT";
  vx_ = 0.0;
  wz_ = 0.0;

  // 仿真位姿初始化（用于里程计反馈）
  sim_x_ = 0.0;
  sim_y_ = 0.0;
  sim_yaw_ = 0.0;

  // 发布器
  state_pub_ =
      this->create_publisher<puppy_interfaces::msg::PlatformMotionState>(
          "/platform/motion_state", 10);

  // 订阅器
  cmd_vel_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
      "/cmd_vel_safe", 10,
      std::bind(&FakeMotionServerNode::onCmdVel, this, std::placeholders::_1));
  posture_sub_ = this->create_subscription<std_msgs::msg::String>(
      "/robot/posture_cmd", 10,
      std::bind(&FakeMotionServerNode::onPosture, this, std::placeholders::_1));

  // 仿真定时器（20Hz）
  sim_timer_ = this->create_wall_timer(
      std::chrono::milliseconds(50),
      std::bind(&FakeMotionServerNode::updateSim, this));

  // 2 秒后自动起立（仿真启动序列）
  auto_stand_timer_ = this->create_wall_timer(
      std::chrono::seconds(2),
      std::bind(&FakeMotionServerNode::autoStand, this));

  RCLCPP_INFO(this->get_logger(),
              "Fake motion server started (will auto-stand in 2s)");
}

void FakeMotionServerNode::autoStand() {
  // 模拟自动起立序列
  if (!standing_) {
    standing_ = true;
    controllable_ = true;
    current_posture_ = "STAND";
    RCLCPP_INFO(this->get_logger(),
                "Fake platform: STOOD UP, controllable=True");
  }
  // 单次触发后取消定时器
  auto_stand_timer_->cancel();
}

void FakeMotionServerNode::onCmdVel(
    const geometry_msgs::msg::Twist::SharedPtr msg) {
  if (!controllable_) {
    return;
  }
  vx_ = msg->linear.x;
  wz_ = msg->angular.z;
  moving_ = std::abs(vx_) > 0.01 || std::abs(wz_) > 0.05;
}

void FakeMotionServerNode::onPosture(
    const std_msgs::msg::String::SharedPtr msg) {
  // 转大写
  std::string posture = msg->data;
  for (auto& c : posture) {
    c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
  }
  RCLCPP_INFO(this->get_logger(), "Fake posture: %s", posture.c_str());
  current_posture_ = posture;

  if (posture == "STAND") {
    standing_ = true;
    controllable_ = true;
  } else if (posture == "SIT" || posture == "LIE_DOWN" ||
             posture == "FREEZE") {
    standing_ = false;
    controllable_ = false;
    vx_ = 0.0;
    wz_ = 0.0;
  }
}

void FakeMotionServerNode::updateSim() {
  // 仿真运动并发布状态
  if (controllable_ && moving_) {
    const double dt = 0.05;
    sim_x_ += vx_ * std::cos(sim_yaw_) * dt;
    sim_y_ += vx_ * std::sin(sim_yaw_) * dt;
    sim_yaw_ += wz_ * dt;
    // 归一化 yaw 到 [-pi, pi]
    while (sim_yaw_ > M_PI) {
      sim_yaw_ -= 2.0 * M_PI;
    }
    while (sim_yaw_ < -M_PI) {
      sim_yaw_ += 2.0 * M_PI;
    }
  }

  puppy_interfaces::msg::PlatformMotionState msg;
  msg.header.stamp = this->now();
  msg.standing = standing_;
  msg.moving = moving_;
  msg.controllable = controllable_;
  msg.gait_mode = moving_ ? "trot" : "idle";
  msg.linear_x = vx_;
  msg.linear_y = 0.0;
  msg.angular_z = wz_;
  if (moving_) {
    msg.platform_state = "EXECUTING";
  } else if (controllable_) {
    msg.platform_state = "READY";
  } else {
    msg.platform_state = "DISABLED";
  }
  state_pub_->publish(msg);
}

}  // namespace puppypi_mock

// ===== 节点入口 =====
int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<puppypi_mock::FakeMotionServerNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/puppypi_mock/puppypi_mock/fake_motion_server.py
"""Fake Motion Server: simulates PuppyPi motion platform.

Responds to /cmd_vel and posture commands, simulates motion execution
with realistic delays and publishes PlatformMotionState.
"""
import time
import math
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import String
from puppy_interfaces.msg import PlatformMotionState


class FakeMotionServerNode(Node):
    """Simulated PuppyPi motion platform."""

    def __init__(self):
        super().__init__('fake_motion_server')

        self.standing = False
        self.moving = False
        self.controllable = False
        self.current_posture = 'SIT'
        self.vx = 0.0
        self.wz = 0.0

        # Simulated position (for odom feedback)
        self.sim_x = 0.0
        self.sim_y = 0.0
        self.sim_yaw = 0.0

        # Publishers
        self.state_pub = self.create_publisher(PlatformMotionState, '/platform/motion_state', 10)

        # Subscribers
        self.create_subscription(Twist, '/cmd_vel_safe', self._on_cmd_vel, 10)
        self.create_subscription(String, '/robot/posture_cmd', self._on_posture, 10)

        # Sim timer (20Hz)
        self.create_timer(0.05, self._update_sim)

        # Auto-stand after 2 seconds (simulated startup)
        self.create_timer(2.0, self._auto_stand)

        self.get_logger().info('Fake motion server started (will auto-stand in 2s)')

    def _auto_stand(self):
        """Simulate automatic stand-up sequence."""
        if not self.standing:
            self.standing = True
            self.controllable = True
            self.current_posture = 'STAND'
            self.get_logger().info('Fake platform: STOOD UP, controllable=True')

    def _on_cmd_vel(self, msg: Twist):
        if not self.controllable:
            return
        self.vx = msg.linear.x
        self.wz = msg.angular.z
        self.moving = abs(self.vx) > 0.01 or abs(self.wz) > 0.05

    def _on_posture(self, msg: String):
        posture = msg.data.upper()
        self.get_logger().info(f'Fake posture: {posture}')
        self.current_posture = posture
        if posture == 'STAND':
            self.standing = True
            self.controllable = True
        elif posture in ('SIT', 'LIE_DOWN', 'FREEZE'):
            self.standing = False
            self.controllable = False
            self.vx = 0.0
            self.wz = 0.0

    def _update_sim(self):
        """Simulate motion and publish state."""
        if self.controllable and self.moving:
            dt = 0.05
            self.sim_x += self.vx * math.cos(self.sim_yaw) * dt
            self.sim_y += self.vx * math.sin(self.sim_yaw) * dt
            self.sim_yaw += self.wz * dt
            # Normalize yaw
            while self.sim_yaw > math.pi:
                self.sim_yaw -= 2 * math.pi
            while self.sim_yaw < -math.pi:
                self.sim_yaw += 2 * math.pi

        msg = PlatformMotionState()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.standing = self.standing
        msg.moving = self.moving
        msg.controllable = self.controllable
        msg.gait_mode = 'trot' if self.moving else 'idle'
        msg.linear_x = self.vx
        msg.linear_y = 0.0
        msg.angular_z = self.wz
        msg.platform_state = 'EXECUTING' if self.moving else ('READY' if self.controllable else 'DISABLED')
        self.state_pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = FakeMotionServerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()
